ons-spark/spark-app/spark_application_and_ui.md

```diff
@@ -0,0 +1,429 @@
+
+Finally, since its release of version 3.0.0, Spark's online documentation contains information about the [Spark web UI](https://spark.apache.org/docs/latest/web-ui.html). Note however that this is a later version of Spark than what is currently used in DAP at the ONS. Therefore there will be some small differences between what you see in your application's UI and that documented in the above link. 
+
+The documentation runs through the different pages of the Spark UI with screenshots and a brief description of the various elements. It's a good place to start and useful for finding definitions. For example, if you want to know what *Scheduler delay* means, search for this term on the docs page and you will find, 
+
+>**Scheduler delay** is the time the task waits to be scheduled for execution
+
+### Moving data outside Spark
+
+Note that in our case using the Data Access Platform (DAP) we could call the *driver* here *CDSW session*, but we'll use driver to be consistent with other material in this book.
+
+What can the Spark UI tell us about the processing of Pandas/R DataFrames? Let's investigate by 
+1. aggregate the `rescue` DataFrame
+2. use `.toPandas()`/`collect()` to move the data to the driver
+3. make a plot of the aggregated data
+4. check the Spark UI
+
+First step is to aggregate the data, we'll find the count of incidents in each `cost_group`. We will distinguish between the Spark and Pandas/R DataFrames using `_spark` or `_pandas`/`_r` as suffixes respectively.
+````{tabs}
+```{code-tab} py
+aggregated_spark = rescue.groupBy("cost_group").agg(F.count("incident_number").alias("count"))
+
+aggregated_pandas = aggregated_spark.toPandas()
+
+aggregated_pandas
+```
+
+```{code-tab} r R
+
+aggregated_spark <- rescue %>% 
+    dplyr::group_by(cost_group) %>%
+    dplyr::summarise(count=n())
+
+aggregated_r <- aggregated_spark %>% sparklyr::collect()
+
+aggregated_r
+
+```
+````
+
+```plaintext
+  cost_group  count
+0       None     82
+1     medium   2045
+2      small   3691
+3      large     80
+```
+
+````{tabs}
+```{code-tab} py
+aggregated_pandas_plot = (
+    aggregated_pandas.fillna("missing")
+    .set_index("cost_group")
+    .loc[["large", "medium", "small", "missing"]]
+)
+
+aggregated_pandas_plot.plot(kind="bar")
+```
+
+```{code-tab} r R
+
+aggregated_r_plot <- aggregated_r %>% dplyr::arrange(desc(cost_group))
+
+ggplot2::ggplot(aggregated_r_plot, ggplot2::aes(cost_group, count)) +
+    ggplot2::geom_bar(stat="identity")
+
+```
+````
+
+```plaintext
+<matplotlib.axes._subplots.AxesSubplot at 0x7faa4841f8d0>
+```
+![Incident cost by cost group chart](../images/spark_app_incident_cost_chart.png)
+
+Now we have our chart, let's see how that translates to tasks in the Spark UI.
+
+There is a job for `.toPandas()`/`collect()`, but nothing after that for the plot. Why? 
+
+The plotting was done in Pandas/R and so Spark was not involved at all. We therefore don't expect the Spark UI to show anything that represents those processes.
```

Comment:
Possibly mention the driver here: The plotting was done in Pandas/R in the driver, and so Spark was not involved at all

Reply:
I did have that in, but because this is a local session, Spark is also executing processes in the driver. So I thought it was a bit confusing